File: Array/3066.Minimum_Operations_to_Exceed_Threshold_Value_II/C++/3066.Minimum_Operations_to_Exceed_Threshold_Value_II.cpp
class Solution {
public:
    int minOperations(vector<int> & nums, int k) 
    {
        //=====================================//
        // Input type:                         //
        //   - vector<int> ref. (nums vector)  //
        //   - int (target threshold)          //
        // Return type:                        //
        //   - int (result minimun operations) //
        //=====================================//

        /*Initialize*/
        ///// Record heap, priority-queue /////
        priority_queue<long long, vector<long long>, greater<long long>> record_pq(begin(nums), end(nums));

        ///// Record minimun values (i.e. 1st, 2nd) /////
        long long record_min_val1, record_min_val2;

        ///// Result minimun operations /////
        int res_min_oprs = 0;


        /*Heap, Priority-queue based loop traversal*/
        while (record_pq.size() >= 2) //Part
        {
            record_min_val1 = record_pq.top(); //Record minimun value (i.e. 1st)

            ///// Check if the current minimun value exceed threshold or not /////
            if (record_min_val1 >= k) { break; }

            record_pq.pop(); //Keep updating, popping

            record_min_val2 = record_pq.top(); //Record minimun value (i.e. 2nd)

            record_pq.pop(); //Keep updating, popping

            record_pq.push(((min(record_min_val1, record_min_val2) * 2) + max(record_min_val1, record_min_val2))); //Keep updating, recording

            (res_min_oprs++); //Keep updating, accumulating

        } //Part

        return res_min_oprs;
    }
};

File: Array/3066.Minimum_Operations_to_Exceed_Threshold_Value_II/Python/3066.Minimum_Operations_to_Exceed_Threshold_Value_II.py
class Solution(object):
    def minOperations(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        #============================================#
        # Heap/Priority-queue based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Update/Heapify nums array #####
        heapify(nums)

        ##### Result minimun operations #####
        res_min_oprs = 0


        ##########################################
        #Heap/Priority-queue based loop traversal
        while (len(nums) >= 2):
            record_min1 = heappop(nums) #Record minimun value-1

            ##### Check if the current minimun value matched conditions or not #####
            if (record_min1 >= k):
                break

            record_min2 = heappop(nums) #Record minimun value-2

            heappush(nums, ((min(record_min1, record_min2) * 2) + max(record_min1, record_min2))) #Keep updating/recording

            res_min_oprs += 1 #Keep updating/accumulating

        return res_min_oprs
